runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/StreamingEngineClientTest.java

```diff
@@ -346,32 +360,33 @@ public void testOnNewWorkerMetadata_redistributesBudget() throws InterruptedExce
             .putAllGlobalDataEndpoints(DEFAULT)
             .build();
 
+    List<WorkerMetadataResponse> workerMetadataResponses =
+        Lists.newArrayList(firstWorkerMetadata, secondWorkerMetadata, thirdWorkerMetadata);
+
+    TestGetWorkBudgetDistributor getWorkBudgetDistributor =
+        spy(new TestGetWorkBudgetDistributor(workerMetadataResponses.size()));
+    streamingEngineClient =
+        newStreamingEngineClient(
+            GetWorkBudget.builder().setItems(1).setBytes(1).build(),
+            getWorkBudgetDistributor,
+            noOpProcessWorkItemFn());
+
     getWorkerMetadataReady.await();
-    fakeGetWorkerMetadataStub.injectWorkerMetadata(firstWorkerMetadata);
-    Thread.sleep(50);
-    fakeGetWorkerMetadataStub.injectWorkerMetadata(secondWorkerMetadata);
-    Thread.sleep(50);
-    fakeGetWorkerMetadataStub.injectWorkerMetadata(thirdWorkerMetadata);
-    Thread.sleep(50);
-    verify(getWorkBudgetDistributor, atLeast(3)).distributeBudget(any(), any());
+
+    // Make sure we are injecting the metadata from smallest to largest.
+    workerMetadataResponses.stream()
+        .sorted(Comparator.comparingLong(WorkerMetadataResponse::getMetadataVersion))
+        .forEach(fakeGetWorkerMetadataStub::injectWorkerMetadata);
+
+    waitForWorkerMetadataToBeConsumed(getWorkBudgetDistributor);
+    verify(getWorkBudgetDistributor, atLeast(workerMetadataResponses.size()))
+        .distributeBudget(any(), any());
   }
 
-  private StreamingEngineConnectionState waitForWorkerMetadataToBeConsumed(
-      int expectedMetadataConsumed) throws InterruptedException {
-    int currentMetadataConsumed = 0;
-    StreamingEngineConnectionState currentConsumedMetadata = StreamingEngineConnectionState.EMPTY;
-    while (true) {
-      if (!connections.get().equals(currentConsumedMetadata)) {
-        ++currentMetadataConsumed;
-        if (currentMetadataConsumed == expectedMetadataConsumed) {
-          break;
-        }
-        currentConsumedMetadata = connections.get();
-      }
-    }
-    // Wait for metadata to be consumed and budgets to be redistributed.
-    Thread.sleep(GetWorkBudgetRefresher.SCHEDULED_BUDGET_REFRESH_MILLIS);
-    return connections.get();
+  private void waitForWorkerMetadataToBeConsumed(
+      TestGetWorkBudgetDistributor getWorkBudgetDistributor) throws InterruptedException {
+    getWorkBudgetDistributor.waitForBudgetDistribution();
+    Thread.sleep(100);
```

Comment:
I just meant to have a sleep in the loop checking the atomic value.
Since we removed that, the sleep is unnecessary

Reply:
gotcha removed